transport/tchannel/channel_transport.go

```diff
@@ -0,0 +1,138 @@
+	"github.com/uber/tchannel-go"
+)
+
+// NewChannelTransport creates a ChannelTransport, suitable for creating inbounds
+// and outbounds with an existing, shared channel.
+//
+// The ChannelTransport uses the underlying TChannel load balancing and peer
+// management, so it is not suitable for use with a peer.Chooser.
+// A future version of YARPC will add a NewTransport constructor that returns
+// a transport suitable for custom peer selection.
+func NewChannelTransport(opts ...TransportOption) *ChannelTransport {
+	var config transportConfig
+	config.tracer = opentracing.GlobalTracer()
+	for _, opt := range opts {
+		opt(&config)
+	}
+
+	// Attempt to construct a channel on behalf of the caller if none given.
+	// Defer the error until Start since NewChannelTransport does not have
+	// an error return.
+	var err error
+	ch := config.ch
+	if config.ch == nil {
+		if config.name == "" {
+			err = fmt.Errorf("can't instantiate TChannelChannelTransport without channel or service name option")
+		} else {
+			ch, err = tchannel.NewChannel(config.name, &tchannel.ChannelOptions{
+				Tracer: config.tracer,
+			})
+		}
+	}
+
+	return &ChannelTransport{
+		ch:     ch,
+		err:    err,
+		addr:   config.addr,
+		tracer: config.tracer,
+	}
+}
+
+// ChannelTransport maintains TChannel peers and creates inbounds and outbounds for
+// TChannel.
+//
+// In a future version, the channel will be suitable for managing peers in a
+// peer.List or other peer.Chooser.
+type ChannelTransport struct {
+	ch     Channel
+	name   string
+	err    error
+	addr   string
+	tracer opentracing.Tracer
+}
+
+// Channel returns the underlying TChannel "Channel" instance.
+func (t *ChannelTransport) Channel() Channel {
+	return t.ch
+}
+
+// ListenAddr exposes the listen address of the transport.
+func (t *ChannelTransport) ListenAddr() string {
+	return t.addr
+}
+
+// Start starts a TChannel transport, opening listening sockets and accepting
+// inbound requests, and opening connections to retained peers.
+//
+// All inbounds must have been assigned a registry to accept inbound requests.
+func (t *ChannelTransport) Start() error {
+
+	// Return error deferred from constructor for the construction of a TChannel.
+	if t.err != nil {
+		return t.err
+	}
+
+	if t.ch.State() == tchannel.ChannelListening {
+		// Channel.Start() was called before RPC.Start(). We still want to
+		// update the Handler and what t.addr means, but nothing else.
+		t.addr = t.ch.PeerInfo().HostPort
+		return nil
+	}
+
+	// Default to ListenIP if addr wasn't given.
+	addr := t.addr
+	if addr == "" {
+		listenIP, err := tchannel.ListenIP()
+		if err != nil {
+			return err
+		}
+
+		addr = listenIP.String() + ":0"
+		// TODO(abg): Find a way to export this to users
+	}
+
+	// TODO(abg): If addr was just the port (":4040"), we want to use
+	// ListenIP() + ":4040" rather than just ":4040".
+
+	if err := t.ch.ListenAndServe(addr); err != nil {
+		return err
+	}
+
+	t.addr = t.ch.PeerInfo().HostPort
```

Comment:
Something to consider in a follow-up. This mechanically extracted from the old code.